ghrn/ghrn.go

```diff
@@ -13,96 +13,118 @@ import (
+	GitHubToken string
+
+	// StopAt is the number of the Pull Request to stop at.
+	// Useful for building the notes of PRs since the last release, for example.
+	StopAt int
+	// IncludeCommits will include commmits messages for each PR.
+	IncludeCommits bool
+	// StopAtLatestRelease will stop at latest release commit.
+	StopAtLatestRelease bool
 }
 
 // BuildReleaseNotes lists GitHub Pull Requests and writes formatted release notes
 // to the given writer.
 func BuildReleaseNotes(ctx context.Context, w io.Writer, conf Config) error {
 
-  if conf.Org == "" {
-    return fmt.Errorf("Config.Org is required")
-  }
-  if conf.Repo == "" {
-    return fmt.Errorf("Config.Repo is required")
-  }
-
-  var httpClient *http.Client
-  if conf.GitHubToken != "" {
-    ts := oauth2.StaticTokenSource(
-      &oauth2.Token{AccessToken: conf.GitHubToken},
-    )
-    httpClient = oauth2.NewClient(ctx, ts)
-  }
+	if conf.Org == "" {
+		return fmt.Errorf("Config.Org is required")
+	}
+	if conf.Repo == "" {
+		return fmt.Errorf("Config.Repo is required")
+	}
+
+	var httpClient *http.Client
+	if conf.GitHubToken != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: conf.GitHubToken},
+		)
+		httpClient = oauth2.NewClient(ctx, ts)
+	}
 	cl := github.NewClient(httpClient)
 
 	opt := &github.PullRequestListOptions{
 		ListOptions: github.ListOptions{PerPage: 100},
 		State:       "closed",
 	}
 
+	repo, _, err := cl.Repositories.Get(ctx, conf.Org, conf.Repo)
+	if err != nil {
+		return fmt.Errorf("get repository: %+v", err)
+	}
+
+	rls, _, err := cl.Repositories.GetLatestRelease(ctx, conf.Org, conf.Repo)
+	if err != nil {
+		return fmt.Errorf("get latest release: %+v", err)
+	}
+
+	comp, _, err := cl.Repositories.CompareCommits(ctx, conf.Org, conf.Repo, rls.GetTagName(), repo.GetDefaultBranch())
+	if err != nil {
+		return fmt.Errorf("compare commitse: %s..%s %+v", rls.GetTagName(), repo.GetDefaultBranch(), err)
+	}
+
+	// new commits from latest release to default branch
+	var newCommits []string
+	for _, commit := range comp.Commits {
+		newCommits = append(newCommits, commit.GetCommit().GetTree().GetSHA())
+	}
+
 	// Iterate over all PRs
 	for {
 		prs, resp, err := cl.PullRequests.List(ctx, conf.Org, conf.Repo, opt)
 		if err != nil {
-      return fmt.Errorf("listing PRs: %s", err)
+			return fmt.Errorf("listing PRs: %s", err)
 		}
 
 		// Iterate over PRs in this page.
 		for _, pr := range prs {
 			if *pr.Number == conf.StopAt {
-        return nil
+				return nil
 			}
 			if pr.MergedAt == nil {
 				continue
 			}
 
+			commits, err := commitsAll(ctx, cl, conf.Org, conf.Repo, pr.GetNumber())
+			if err != nil {
+				return fmt.Errorf("listing PR commits: %s", err)
+			}
+			var prCommits []string
+			for _, commit := range commits {
+				prCommits = append(prCommits, commit.GetCommit().GetTree().GetSHA())
+			}
+
+			if conf.StopAtLatestRelease && !any(prCommits, newCommits) {
+				// stop any new commits do not contains pr commits
```

Comment:
A little help with the English: "Stop when a PR doesn't contain any commits from since the latest release."

Reply:
Also, technically it's possible to have a PR that was merged to a branch besides master after the latest release, in which case this code would stop early and produce incorrect release notes.